src/Coupon/WCCouponAdapter.php

```diff
@@ -0,0 +1,277 @@
+
+    /**
+     * Initialize this object's properties from an array.
+     *
+     * @param array $array
+     *            Used to seed this object's properties.
+     *            
+     * @return void
+     *
+     * @throws InvalidValue When a WooCommerce coupon is not provided or it is invalid.
+     */
+    public function mapTypes( $array ) {
+        if ( 
+            empty( $array['wc_coupon'] ) ||
+            ! $array['wc_coupon'] instanceof WC_Coupon ) {
+            throw InvalidValue::not_instance_of( WC_Coupon::class, 'wc_coupon' );
+        }
+
+        $wc_coupon = $array['wc_coupon'];
+        $this->wc_coupon_id = $wc_coupon->get_id();
+        $this->map_woocommerce_coupon( $wc_coupon );
+
+        // Google doesn't expect extra fields, so it's best to remove them
+        unset( $array['wc_coupon'] );
+
+        parent::mapTypes( $array );
+    }
+
+    /**
+     * Map the WooCommerce coupon attributes to the current class.
+     *
+     * @return void
+     */
+    protected function map_woocommerce_coupon( WC_Coupon $wc_coupon ) {
+        $this->setRedemptionChannel( self::CHANNEL_ONLINE );
+
+        $content_language = empty( get_locale() ) ? 'en' : strtolower( 
+            substr( get_locale(), 0, 2 ) ); // ISO 639-1.
+        $this->setContentLanguage( $content_language );
+
+        $this->map_wc_coupon_id( $wc_coupon )
+            ->map_wc_general_attributes( $wc_coupon )
+            ->map_wc_usage_restriction( $wc_coupon );
+    }
+
+    /**
+     * Map the WooCommerce coupon ID.
+     *
+     * @return $this
+     */
+    protected function map_wc_coupon_id( WC_Coupon $wc_coupon ): WCCouponAdapter {
+        $coupon_id = "{$this->get_slug()}_{$wc_coupon->get_id()}";
+        $this->setPromotionId( $coupon_id );
+
+        return $this;
+    }
+
+    /**
+     * Map the general WooCommerce coupon attributes.
+     *
+     * @return $this
+     */
+    protected function map_wc_general_attributes( WC_Coupon $wc_coupon ): WCCouponAdapter {
+        $this->setOfferType( self::OFFER_TYPE_GENERIC_CODE );
+        $this->setGenericRedemptionCode( $wc_coupon->get_code() );
+
+        $coupon_amount = $wc_coupon->get_amount();
+        if ( $wc_coupon->is_type( self::WC_DISCOUNT_TYPE_PERCENT ) ) {
+            $this->setCouponValueType( self::COUPON_VALUE_TYPE_PERCENT_OFF );
+            $percent_off = round( $coupon_amount );
+            $this->setPercentOff( $percent_off );
+            $this->setLongtitle( sprintf( '%d off', $percent_off ) );
+        } else if ( $wc_coupon->is_type( 
+            [
+                self::WC_DISCOUNT_TYPE_FIXED_CART,
+                self::WC_DISCOUNT_TYPE_FIXED_PRODUCT] ) ) {
+            $this->setCouponValueType( self::COUPON_VALUE_TYPE_MONEY_OFF );
+            $this->setMoneyOffAmount( 
+                $this->map_google_price_amount( $coupon_amount ) );
+            $this->setLongtitle( 
+                sprintf( 
+                    '%d %s% off',
```

Comment:
For clarity, we should escape the literal `%` character.

```suggestion
                    '%d %s%% off',
```

Reply:
Updated.